src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -345,15 +409,35 @@ public void OnResponseAbort(Exception abortException)
                 }
             }
 
-            public async Task ReadResponseHeadersAsync()
+            public async Task ReadResponseHeadersAsync(TaskCompletionSource<bool> shouldSendRequestBody = null)
             {
                 // Wait for response headers to be read.
-                (bool wait, bool emptyResponse) = TryEnsureHeaders();
-                if (wait)
+                bool emptyResponse;
+                bool wait;
+
+                // Process all informational responses if any and wait for final status.
+                do
                 {
-                    await GetWaiterTask().ConfigureAwait(false);
                     (wait, emptyResponse) = TryEnsureHeaders();
-                    Debug.Assert(!wait);
+                    if (wait)
+                    {
+                        await GetWaiterTask().ConfigureAwait(false);
+                        (wait, emptyResponse) = TryEnsureHeaders();
+                    }
+
+                    if (Response.StatusCode == HttpStatusCode.Continue)
+                    {
+                        if (NetEventSource.IsEnabled) _connection.Trace("Received 100Continue status.");
+                        shouldSendRequestBody?.TrySetResult(true);
+                        shouldSendRequestBody = null;
+                    }
+                }
+                while ((int)Response.StatusCode < 200);
+
+                if ((int)Response.StatusCode >= 300)
+                {
+                    // If we tried 100-Continue and got rejected signal that we should not send request body.
+                    shouldSendRequestBody?.TrySetResult(false);
```

Comment:
Shouldn't we also complete the task (set to true) if the response is < 300? E.g. 200? As it stands, it looks like we'll wait for the timer in this case...

Reply:
original assumption was that if we get final response we are already sending body e.g. timer expired. It is probably cheap so I can do trySetResult always and set value based on result code.

Reply:
Yeah, that's probably true, but just in case, it seems reasonable to set the task here anyway.